worldedit-forge/src/main/java/com/sk89q/worldedit/forge/mixin/MixinLevelChunkSetBlockHook.java

```diff
@@ -0,0 +1,85 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.forge.mixin;
+
+import com.sk89q.worldedit.forge.internal.ExtendedChunk;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Registry;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.LevelHeightAccessor;
+import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.chunk.LevelChunkSection;
+import net.minecraft.world.level.chunk.UpgradeData;
+import net.minecraft.world.level.levelgen.blending.BlendingData;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.Slice;
+
+import javax.annotation.Nullable;
+
+@Mixin(LevelChunk.class)
+public abstract class MixinLevelChunkSetBlockHook extends ChunkAccess implements ExtendedChunk {
+    private boolean shouldUpdate = true;
+
+    public MixinLevelChunkSetBlockHook(ChunkPos chunkPos, UpgradeData upgradeData, LevelHeightAccessor levelHeightAccessor, Registry<Biome> registry, long l, @org.jetbrains.annotations.Nullable LevelChunkSection[] levelChunkSections, @org.jetbrains.annotations.Nullable BlendingData blendingData) {
+        super(chunkPos, upgradeData, levelHeightAccessor, registry, l, levelChunkSections, blendingData);
+    }
+
+    @Nullable
+    @Override
+    public BlockState setBlockState(BlockPos pos, BlockState state, boolean moved, boolean update) {
+        // save the state for the hook
+        shouldUpdate = update;
+        try {
+            return setBlockState(pos, state, moved);
+        } finally {
+            // restore natural mode
+            shouldUpdate = true;
+        }
+    }
+
+    @Redirect(
+        method = "setBlockState",
+        slice = @Slice(
+            from = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/block/state/BlockState;is(Lnet/minecraft/world/level/block/Block;)Z")
+        ),
+        at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/block/state/BlockState;onPlace(Lnet/minecraft/world/level/Level;Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;Z)V")
+    )
+    public void setBlockStateHook(BlockState target, Level world, BlockPos pos, BlockState old, boolean move) {
+        boolean localShouldUpdate;
+        MinecraftServer server = world.getServer();
+        if (server == null || Thread.currentThread() != server.getRunningThread()) {
```

Comment:
why not just `if (world.isClientSide/world.isRemote/whatever the mappings are)`?

Reply:
This was copied from Fabric - I am assuming there was a reason for it (probably in case something is dumb and calls updates on another thread)

Reply:
Especially because the `shouldUpdate` flag isn't volatile, that would be a thread safety violation.